### ISSUE TITLE
[Request] SteamOS3 Keyboard - please add Up, Down, F1-F12, Alt, Ctrl, Break, Print and Umlaute...

### ISSUE THREAD

**Issue by derpelikan** [ValveSoftware/SteamOS#752]
The new TrackPack functionality is amazing, yet there are still situations the onscreen-keyboard is not sufficiant and complicated SteamInput mappings must be done to just push one button.
Up, Down, F1-F12, Alt, Ctrl, Break, Print and Umlaute are currently unavailable. They don't need to be easily accessible, but there should be some way to reach them from the keyboard, anyway.

**Comment by justbugs97** [comment]
More customization overall would be fantastic, I love that they brought back dual trackpad typing.

**Comment by kiwidoggie** [comment]
I'm surprised they didn't have an option for CTRL at least. Hopefully this gets updated in the near future :)

**Comment by professornope** [comment]
I made a couple of these and common keyboard shortcuts through the controller configuration for Desktop.
If you go into Desktop > Steam Settings > Controller > Desktop Configuration you can map the D pad and buttons to anything. For example, my down is alt, right is win, left is ctrl. Then on the YXAB, A is Tab, X is C, B is V, and Y is Z. Kinda confusing at first but works very well. So if I do Left+X, I get copy, then Left+B to paste. You can map any keyboard button to any controller button, along with mode switching, a lot is available to you already!

**Comment by TOGLK** [comment]
I would love too see the keyboard get these much needed additions, as it currently stands I waste a lot of time trying to enter special characters.

Taking a page out of the way that most Android keyboards handle the limitations of a small screen would be a good start I feel.

Having multiple "pages" of different keyboard layouts that can be accessed using the existing "Arrow" keys on the current layout.

Clicking the "Right" Arrow Key would present you with a Layout Offering a wide range of special characters which if you click-&-hold on the keys other option would appear for you to choose from for things like currency, maths symbols, special formating, accented letters, etc.

Clicking the "Left" Arrow Key would present you with a layout offering special Key Inputs like CTRL, ALT, SHIFT, ALT-GR, META/Super/Windows/Command, Directional Arrow Keys, Pause Break, Num Lock, Scroll Lock, Page Up & Down, End, Home, the full suite on Numpad Inputs.

And if you pressed the "Left" Arrow a second time you would then be given a blank keyboard layout where you could setup your own custom character/symbol inputs to be used with just a single press, for those of us that have perhaps a specific workflow in mind.

The opportunity to improve the keyboard here is huge!

**Comment by werpu** [comment]
Yes the up and down keys are really needed for the usecase mentioned. Why they have omitted the function keys is a little bit beyond me, they are heavily used especially in older games for quicksave/restore quick access and entering the games menu.
One way to solve this would be an extra screen in the keyboard to reach those keys, sort of like you reach the emojis that way (has anyone ever used the emoji screen?)

**Comment by mindinsomnia** [comment]
The up/down keys are perhaps my biggest 'wants' because so often when I'm using the detect for web browsing, I'll be typing in an address or text into a text field, and suggestions will come up below it as auto suggestions. On a regular physical keyboard you would access them by pressing the down key, but on the Deck unless you have a physical button mapped to that at the time, there's no way to access that. You can't press a mapped button on the deck itself to do it either because the controls of the deck are mapped to the OSK while it's focused, pressing left/right/up/down on the dpad just moves around the currently highlighted key on the OSK.

**Comment by kisak-valve** [comment]
Some progress "Added up/down cursor keys to keyboard, press shift then left/right cursor to use." in the [2023-01-16](https://steamcommunity.com/games/1675200/announcements/detail/5662747178724589491) Steam Deck client beta update.

**Comment by Hahasamian** [comment]
Been a year since the last post and it doesn't seem like we have any progress towards this...
I feel like an "fn" key right next to the emoji key makes the most sense. As it is the keyboard looks kinda unbalanced too.

**Comment by cubimon** [comment]
Did someone try to read touch pad data to create a custom keyboard? I tried to read /dev/input/event3, but it seems to be blocked while steam is running, if steam is stopped it could be used to read the touchpad input I guess

**Comment by cheater** [comment]
a full keyboard layout really is needed. sometimes games require you to press a specific button, and when using the desktop, sometimes you just can't go around using specific keys. Could this be implemented? pretty please?

**Comment by cubimon** [comment]
I did implement my own keyboard [here](https://github.com/cubimon/steamdeck-keyboard). But it is still WIP, there are also some other generic virtual keyboards which aren't that great on the steamdeck
